com.unity.perception/Runtime/GroundTruth/Labelers/BoundingBox3DLabeler.cs

```diff
@@ -0,0 +1,278 @@
+            public float[] acceleration; // TODO
+        }
+
+        static ProfilerMarker s_BoundingBoxCallback = new ProfilerMarker("OnBoundingBoxes3DReceived");
+        AnnotationDefinition m_AnnotationDefinition;
+        List<BoxData> m_BoundingBoxValues;
+
+        int m_CurrentFrame;
+
+        /// <inheritdoc/>
+        protected override bool supportsVisualization => false;
+
+        /// <summary>
+        /// Fired when the bounding boxes are computed for a frame.
+        /// </summary>
+        public event Action<int, BoxData> BoundingBoxComputed;
+
+        /// <summary>
+        /// Creates a new BoundingBox3DLabeler. Be sure to assign <see cref="idLabelConfig"/> before adding to a <see cref="PerceptionCamera"/>.
+        /// </summary>
+        public BoundingBox3DLabeler() {}
+
+        /// <summary>
+        /// Creates a new BoundingBox3DLabeler with the given <see cref="IdLabelConfig"/>.
+        /// </summary>
+        /// <param name="labelConfig">The label config for resolving the label for each object.</param>
+        public BoundingBox3DLabeler(IdLabelConfig labelConfig)
+        {
+            this.idLabelConfig = labelConfig;
+        }
+
+        /// <inheritdoc/>
+        protected override void Setup()
+        {
+            if (idLabelConfig == null)
+                throw new InvalidOperationException("BoundingBox2DLabeler's idLabelConfig field must be assigned");
+
+            var updater = World.DefaultGameObjectInjectionWorld?.GetOrCreateSystem<GroundTruthUpdateSystem>();
+            updater?.Activate(this);
+
+            m_AnnotationDefinition = DatasetCapture.RegisterAnnotationDefinition("bounding box 3D", idLabelConfig.GetAnnotationSpecification(),
+                "Bounding box for each labeled object visible to the sensor", id: new Guid(annotationId));
+        }
+
+        /// <inheritdoc/>
+        protected override void Cleanup()
+        {
+            var updater = World.DefaultGameObjectInjectionWorld?.GetExistingSystem<GroundTruthUpdateSystem>();
+            updater?.Deactivate(this);
+        }
+
+        Plane[] m_PerceptionCameraFrustumFrames;
+        
+        /// <inheritdoc/>
+        public void OnBeginUpdate()
+        {
+            if (m_BoundingBoxValues == null)
+                m_BoundingBoxValues = new List<BoxData>();
+            else
+                m_BoundingBoxValues.Clear();
+
+            m_PerceptionCameraFrustumFrames = GeometryUtility.CalculateFrustumPlanes(perceptionCamera.attachedCamera);
+            m_CurrentFrame = Time.frameCount;
+        }
+
+        static BoxData ConvertToBoxData(IdLabelEntry label, uint instanceId, Vector3 center, Vector3 extents, Quaternion rotation)
+        {
+            return new BoxData
+            {
+                label_id = label.id,
+                label_name = label.label,
+                instance_id = instanceId,
+                translation = new[] { center.x, center.y, center.z },
```

Comment:
These should be able to be Vector3 or Quaternions. Try to avoid allocating arrays, as the GC is slow.

Reply:
I'm doing this because Vector3 and Quaternions are not serializable. I'm not sure if we want to add a way to work around this.

Reply:
Hmm, it looks like AnnotationAsync.ReportValues is still using JArray.FromObject() instead of JArrayFromArray. I tried serializing a struct with Vector3 and Quaternion to Json using `DatasetJsonUtility.ToJToken(hasVec);` and it worked fine, so with a little bit of rework of AnnotationAsync.ReportValues you should be able to make the switch.

Reply:
This has been fixed, I found a pretty cool way that we could directly support reading and writing Vector3 and Quaternions to json